stats/thresholds.go

```diff
@@ -26,70 +25,162 @@ import (
 		AbortGracePeriod: gracePeriod,
-		pgm:              pgm,
-		rt:               newThreshold,
 	}, nil
 }
 
-func (t Threshold) runNoTaint() (bool, error) {
-	v, err := t.rt.RunProgram(t.pgm)
-	if err != nil {
-		return false, err
+func (t *Threshold) runNoTaint(sinks map[string]float64) (bool, error) {
+	// Extract the sink value for the aggregation method used in the threshold
+	// expression
+	lhs, ok := sinks[t.parsed.AggregationMethod]
+	if !ok {
+		return false, fmt.Errorf("unable to apply threshold %s over metrics; reason: "+
+			"no metric supporting the %s aggregation method found",
+			t.parsed.AggregationMethod,
+			t.parsed.AggregationMethod)
+	}
+
+	// Apply the threshold expression operator to the left and
+	// right hand side values
+	var passes bool
+	switch t.parsed.Operator {
+	case ">":
+		passes = lhs > t.parsed.Value
+	case ">=":
+		passes = lhs >= t.parsed.Value
+	case "<=":
+		passes = lhs <= t.parsed.Value
+	case "<":
+		passes = lhs < t.parsed.Value
+	case "==":
+		passes = lhs == t.parsed.Value
+	case "===":
+		// Considering a sink always maps to float64 values,
+		// strictly equal is equivalent to loosely equal
+		passes = lhs == t.parsed.Value
+	case "!=":
+		passes = lhs != t.parsed.Value
+	default:
+		// The ParseThresholdCondition constructor should ensure that no invalid
+		// operator gets through, but let's protech our future selves anyhow.
+		return false, fmt.Errorf("unable to apply threshold %s over metrics; "+
+			"reason: %s is an invalid operator",
+			t.Source,
+			t.parsed.Operator,
+		)
 	}
-	return v.ToBoolean(), nil
+
+	// Perform the actual threshold verification
+	return passes, nil
 }
 
-func (t *Threshold) run() (bool, error) {
-	b, err := t.runNoTaint()
-	t.LastFailed = !b
-	return b, err
+func (t *Threshold) run(sinks map[string]float64) (bool, error) {
+	passes, err := t.runNoTaint(sinks)
+	t.LastFailed = !passes
+	return passes, err
+}
+
+type thresholdCondition struct {
+	AggregationMethod string
+	Operator          string
+	Value             float64
+}
+
+// ParseThresholdCondition parses a threshold condition expression,
+// as defined in a JS script (for instance p(95)<1000), into a ThresholdCondition
+// instance, using our parser combinators package.
+
+// This parser expect a threshold expression matching the following BNF
+//
+// ```
+// assertion           -> aggregation_method whitespace* operator whitespace* float
+// aggregation_method  -> trend | rate | gauge | counter
+// counter             -> "count" | "sum" | "rate"
+// gauge               -> "last" | "min" | "max" | "value"
+// rate                -> "rate"
+// trend               -> "min" | "mean" | "avg" | "max" | percentile
+// percentile          -> "p(" float ")"
+// operator            -> ">" | ">=" | "<=" | "<" | "==" | "===" | "!="
+// float               -> digit+ (. digit+)?
+// digit               -> "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9"
+// whitespace          -> space | tab
```

Comment:
That's a good question. It's probably one of the blind spots of the BNF. The parser as I wrote it discards `\n`, `\r`, `\t`, and `\r\n`. I added really as sentinels, to be sure that even if programmers were to add them to their expression, they wouldn't impact the execution. The goal of a BNF being to be as explicit as possible on the format, I would actually add a definition for newline, as being `newline -> "\n" | "\r\n"`, and add it to assertion as optional. Does that address your question? 🐻

Reply:
Here's an updated BNF to explicit what I meant:

```
assertion           -> aggregation_method whitespace* operator whitespace* float newline*
aggregation_method  -> trend | rate | gauge | counter
counter             -> "count" | "sum" | "rate"
gauge               -> "last" | "min" | "max" | "value"
rate                -> "rate"
trend               -> "min" | "mean" | "avg" | "max" | percentile
percentile          -> "p(" float ")"
operator            -> ">" | ">=" | "<=" | "<" | "==" | "===" | "!="
float               -> digit+ (. digit+)?
digit               -> "0" | "1" | "2" | "3" | "4" | "5" | "6" | "7" | "8" | "9"
whitespace          -> space | tab
tab                 -> "\t"
space               -> " "
newline		 -> linefeed | crlf
crlf 		-> carriage_return linefeed
linefeed		-> "\n"
carriage_return	-> "\r"
```

Reply:
Yes, I missed that yesterday: currently tabs are allowed inside threshold expression and newline is allowed in the end as you described.
So this is again more a question of design: _should_ there be tabs or newlines there? What's the use case that someone would need those chars in threshold expression? I was more thinking in line of simplifying it as much as possible and maybe help people avoid some mistakes in writing tests. I guess it can be seen both ways :thinking: 

And I agree that BNF is a very nice way to be clear on expectations in this case :+1:

Reply:
That's a good question. A decision needs to be made here. As with the rest of this PR, my approach has been to get as close as possible to what is documented and already implemented (with my probably limited view of it still). I would tend to think that having the parser discard spaces and tabs is a nice to have, as it makes it permissive to "not perfect" input. I have no opinion on the newlines. Furthermore, I don't think I have the experience/knowledge of the test scripts and how they're used in the wild yet to be an able to decide, so I'll let you folks decide :)

Reply:
if the cloud supports tabs we should support them as well, if not - not

Reply:
Looking at the cloud's parser, they drop spaces, but do not seem to touch at tabs, nor newlines actually: I assume they produce an error in that case.

See https://github.com/grafana/k6-cloud-lib/blob/develop/k6_lib/utils/thresholds.py#L66

Thus I will make the parser *not* ignore tabs nor new lines.